hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1052,51 +1091,81 @@ protected HoodieData<HoodieRecord> prepRecords(Map<MetadataPartitionType,
+   * with smaller timestamps to metadata table (makes for easier debugging)
+   * <p>
+   * This adds the limitations that long-running async operations (clustering, etc.) may cause delay in such MDT
+   * optimizations. We will relax this after MDT code has been hardened.
    */
-  protected void compactIfNecessary(BaseHoodieWriteClient writeClient, String instantTime) {
-    // finish off any pending compactions if any from previous attempt.
-    writeClient.runAnyPendingCompactions();
-
-    String latestDeltaCommitTimeInMetadataTable = metadataMetaClient.reloadActiveTimeline()
-        .getDeltaCommitTimeline()
-        .filterCompletedInstants()
-        .lastInstant().orElseThrow(() -> new HoodieMetadataException("No completed deltacommit in metadata table"))
-        .getTimestamp();
-    // we need to find if there are any inflights in data table timeline before or equal to the latest delta commit in metadata table.
-    // Whenever you want to change this logic, please ensure all below scenarios are considered.
-    // a. There could be a chance that latest delta commit in MDT is committed in MDT, but failed in DT. And so findInstantsBeforeOrEquals() should be employed
-    // b. There could be DT inflights after latest delta commit in MDT and we are ok with it. bcoz, the contract is, latest compaction instant time in MDT represents
-    // any instants before that is already synced with metadata table.
-    // c. Do consider out of order commits. For eg, c4 from DT could complete before c3. and we can't trigger compaction in MDT with c4 as base instant time, until every
-    // instant before c4 is synced with metadata table.
-    List<HoodieInstant> pendingInstants = dataMetaClient.reloadActiveTimeline().filterInflightsAndRequested()
-        .findInstantsBeforeOrEquals(latestDeltaCommitTimeInMetadataTable).getInstants();
+  @Override
+  public void performTableServices(Option<String> inFlightInstantTimestamp) {
+    HoodieTimer metadataTableServicesTimer = HoodieTimer.start();
+    boolean allTableServicesExecutedSuccessfullyOrSkipped = true;
+    try {
+      BaseHoodieWriteClient writeClient = getWriteClient();
+      // Run any pending table services operations.
+      runPendingTableServicesOperations(writeClient);
+
+      // Check and run clean operations.
+      String latestDeltacommitTime = metadataMetaClient.reloadActiveTimeline().getDeltaCommitTimeline()
+              .filterCompletedInstants()
+              .lastInstant().get()
+              .getTimestamp();
+      LOG.info("Latest deltacommit time found is " + latestDeltacommitTime + ", running clean operations.");
+      cleanIfNecessary(writeClient, latestDeltacommitTime);
+
+      // Do timeline validation before scheduling compaction/logcompaction operations.
+      if (!validateTimelineBeforeSchedulingCompaction(inFlightInstantTimestamp, latestDeltacommitTime)) {
+        return;
```

Comment:
We should not return directly because the archiving is also blocked, if no compaction plan should be scheduled, the archiving should also be triggered.

Reply:
I see your point. but in reality, I feel we may not gain much. unless compaction in MDT kicks in, archival might not have anything to do after last time it was able to archive something. So, not a bad idea to not trigger archival if there is no compaction. 
Same applies for clean as well. but lets not optimize anything in this patch.

Reply:
> unless compaction in MDT kicks in, archival might not have anything to do after last time it was able to archive something.

Then archiving will always be blocked by the compaction.